client/blockchain-service/src/types.rs

```diff
@@ -128,3 +127,93 @@ pub enum ExtrinsicResult {
 
 /// Type alias for the extrinsic hash.
 pub type ExtrinsicHash = H256;
+
+/// Type alias for the tip.
+pub type Tip = pallet_transaction_payment::ChargeTransactionPayment<storage_hub_runtime::Runtime>;
+
+/// A struct which defines a submit extrinsic retry strategy.
+pub struct RetryStrategy {
+    /// Maximum number of retries after which the extrinsic submission will be considered failed.
+    pub max_retries: u32,
+    /// Maximum time to wait for a response before assuming the extrinsic submission has failed.
+    pub timeout: Duration,
+    /// Maximum tip to be paid for the extrinsic submission. The progression follows an exponential
+    /// backoff strategy.
+    pub max_tip: f64,
+    /// Base multiplier for the tip calculation.
+    /// This is a constant value that is used to calculate the tip multiplier.
+    /// A higher value will make tips grow faster.
+    pub base_multiplier: f64,
+    /// An optional check function to determine if the extrinsic should be retried.
+    pub should_retry: Option<Box<dyn Fn() -> Pin<Box<dyn Future<Output = bool> + Send>> + Send>>,
+}
+
+impl RetryStrategy {
+    /// Creates a new `RetryStrategy` instance.
+    pub fn new(max_retries: u32, timeout: Duration, max_tip: f64, base_multiplier: f64) -> Self {
+        Self {
+            max_retries,
+            timeout,
+            max_tip,
+            base_multiplier,
+            should_retry: None,
+        }
+    }
+
+    pub fn with_max_retries(mut self, max_retries: u32) -> Self {
+        self.max_retries = max_retries;
+        self
+    }
+
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = timeout;
+        self
+    }
+
+    pub fn with_max_tip(mut self, max_tip: f64) -> Self {
+        self.max_tip = max_tip;
+        self
+    }
+
+    pub fn with_base_multiplier(mut self, base_multiplier: f64) -> Self {
+        self.base_multiplier = base_multiplier;
+        self
+    }
+
+    pub fn with_should_retry(
+        mut self,
+        should_retry: Option<Box<dyn Fn() -> Pin<Box<dyn Future<Output = bool> + Send>> + Send>>,
+    ) -> Self {
+        self.should_retry = should_retry;
+        self
+    }
+
+    /// Compute the exponential increase (multiplier) in tip at each retry.
+    /// A higher multiplier will make tips grow exponentially faster.
+    fn compute_tip_multiplier(&self) -> f64 {
+        (self.base_multiplier.ln() / self.max_retries as f64).exp()
+    }
+
+    pub fn compute_tip(&self, retry_count: u32) -> f64 {
+        let multiplier = self.compute_tip_multiplier();
+
+        // Calculate the geometric progression factor for each retry
+        let factor = (multiplier.powf(retry_count as f64) - 1.0)
+            / (multiplier.powf(self.max_retries as f64) - 1.0);
```

Comment:
Optimisation:
```rust
multiplier.powf(self.max_retries as f64)
```

Is just `base_multiplier`

Reply:
Is it ok that this function doesn't guard against `retry_count` being grater than `max_retires`?